File: src/main/java/akostenko/aicars/menu/MenuState.java
package akostenko.aicars.menu;

import static org.lwjgl.input.Keyboard.KEY_DOWN;
import static org.lwjgl.input.Keyboard.KEY_LEFT;
import static org.lwjgl.input.Keyboard.KEY_RETURN;
import static org.lwjgl.input.Keyboard.KEY_RIGHT;
import static org.lwjgl.input.Keyboard.KEY_UP;

import akostenko.aicars.GameStateIds;
import akostenko.aicars.Game;
import akostenko.aicars.GameSettings;
import akostenko.aicars.keyboard.KeyboardHelper;
import akostenko.aicars.keyboard.SingleKeyAction;
import org.newdawn.slick.Color;
import org.newdawn.slick.GameContainer;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.Input;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.TrueTypeFont;
import org.newdawn.slick.state.BasicGameState;
import org.newdawn.slick.state.StateBasedGame;

import java.awt.Font;
import java.util.Arrays;
import java.util.List;

public class MenuState extends BasicGameState{

    //////////////////////////////////////////
    //   |     |       20%              |   //
    //---+------------------------------+---//
    //   |SubM1  item1  item2           |   //
    //   |                              |   //
    //20%|SubM2  item3  item4           |20%//
    //   |                              |   //
    //   |SubMN                         |   //
    //---+------------------------------+---//
    //   |             20%              |   //
    //////////////////////////////////////////


    private final StartButton startButton = new StartButton();
    private final ModeMenu modeMenu = new ModeMenu();
    private final TrackMenu trackMenu = new TrackMenu();
    private final List<SubMenu<?>> menu = Arrays.asList(startButton, modeMenu, trackMenu);

    private int currentMenu = 0;

    private final int lineWidth = 3;
    private final Color fontColor = new Color(220, 200, 50);
    private final float leftMargin = Game.WIDTH * 0.2f;
    private final float rightMargin = Game.WIDTH * 0.2f;
    private final float topMargin = Game.HEIGHT * 0.2f;
    private float bottomMargin = Game.HEIGHT * 0.2f;
    private final int subMenuHeight = 36;
    private final int subMenuItemHeight = 24;
    private final float subMenuWidth = Game.WIDTH * 0.2f;
    private final TrueTypeFont submenuFont = new TrueTypeFont(new Font(Font.SANS_SERIF, Font.BOLD, subMenuHeight), true);
    private final TrueTypeFont itemFont = new TrueTypeFont(new Font(Font.SANS_SERIF, Font.BOLD, subMenuItemHeight), true);

    @Override
    public int getID() {
        return GameStateIds.getId(this.getClass());
    }

    @Override
    public void init(GameContainer container, StateBasedGame game) throws SlickException {
        Input input = container.getInput();
        GameSettings.get().getGlobalListeners().forEach(input::addKeyListener);
        KeyboardHelper.getKeyListeners().forEach(input::addKeyListener);
        input.addKeyListener(new SingleKeyAction(v -> menuChange(-1), KEY_UP));
        input.addKeyListener(new SingleKeyAction(v -> menuChange(+1), KEY_DOWN));

        input.addKeyListener(new SingleKeyAction(v -> selectionChange(-1), KEY_LEFT));
        input.addKeyListener(new SingleKeyAction(v -> selectionChange(+1), KEY_RIGHT));

        input.addKeyListener(new SingleKeyAction(v -> enterMenu(), KEY_RETURN));

        container.setTargetFrameRate(60);
    }

    private void menuChange(int delta) {
        currentMenu +=delta + menu.size();
        currentMenu %= menu.size();
    }

    private void selectionChange(int delta) {
        menu.get(currentMenu).change(delta);
    }

    private void enterMenu() {
        menu.get(currentMenu).enter();
    }

    @Override
    public void render(GameContainer container, StateBasedGame game, Graphics g) throws SlickException {
        g.setLineWidth(lineWidth);
        g.setColor(fontColor);

        menu.forEach(submenu -> {
            float currentY = topMargin + menu.indexOf(submenu) * (subMenuHeight + 3 * lineWidth);
            drawSubMenu(g, submenu, currentY);

            drawSubmenuItems(g, submenu, currentY);
        });
    }

    private void drawSubMenu(Graphics g, SubMenu<?> submenu, float currentY) {
        g.setFont(submenuFont);
        g.drawString(submenu.getTitle(), leftMargin, currentY);
        if (currentMenu == menu.indexOf(submenu)) {
            g.drawLine(leftMargin, currentY + subMenuHeight + lineWidth, leftMargin + subMenuWidth - 5*lineWidth, currentY + subMenuHeight + lineWidth);
        }
    }

    private void drawSubmenuItems(Graphics g, SubMenu<?> submenu, float currentY) {
        g.setFont(itemFont);
        float itemY = currentY + subMenuHeight - subMenuItemHeight;
        List<? extends MenuItem> items = submenu.getItems();
        float itemWidth = (Game.WIDTH - leftMargin - rightMargin - subMenuWidth) / items.size();
        items.forEach(item -> {
            float currentX = leftMargin + subMenuWidth + submenu.getItems().indexOf(item) * itemWidth;
            g.drawString(item.getTitle(), currentX, itemY);
            if (submenu.isCurrent(item)) {
                g.drawLine(currentX, itemY + subMenuItemHeight + lineWidth, currentX + itemWidth - 5*lineWidth, itemY + subMenuItemHeight + lineWidth);
            }
        });
    }

    @Override
    public void update(GameContainer container, StateBasedGame game, int delta) throws SlickException {

    }

    @Override
    public void leave(GameContainer container, StateBasedGame game) throws SlickException {
        super.leave(container, game);
        GameSettings.get()
                .setTrack(trackMenu.getCurrent())
                .setMode(modeMenu.getCurrent())
                .save();
    }

    @Override
    public void enter(GameContainer container, StateBasedGame game) throws SlickException {
        super.enter(container, game);
        trackMenu.setCurrent(GameSettings.get().getTrack());
        modeMenu.setCurrent(GameSettings.get().getMode());
    }


}

File: src/main/java/akostenko/aicars/menu/MenuState.kt
package akostenko.aicars.menu

import akostenko.aicars.Game
import akostenko.aicars.GameSettings
import akostenko.aicars.GameStateIds
import akostenko.aicars.keyboard.KeyboardHelper
import akostenko.aicars.keyboard.SingleKeyAction
import org.lwjgl.input.Keyboard.KEY_DOWN
import org.lwjgl.input.Keyboard.KEY_LEFT
import org.lwjgl.input.Keyboard.KEY_RETURN
import org.lwjgl.input.Keyboard.KEY_RIGHT
import org.lwjgl.input.Keyboard.KEY_UP
import org.newdawn.slick.Color
import org.newdawn.slick.GameContainer
import org.newdawn.slick.Graphics
import org.newdawn.slick.KeyListener
import org.newdawn.slick.SlickException
import org.newdawn.slick.TrueTypeFont
import org.newdawn.slick.state.BasicGameState
import org.newdawn.slick.state.StateBasedGame
import java.awt.Font

class MenuState : BasicGameState() {

    //////////////////////////////////////////
    //   |     |       20%              |   //
    //---+------------------------------+---//
    //   |SubM1  item1  item2           |   //
    //   |                              |   //
    //20%|SubM2  item3  item4           |20%//
    //   |                              |   //
    //   |SubMN                         |   //
    //---+------------------------------+---//
    //   |             20%              |   //
    //////////////////////////////////////////


    private val startButton = StartButton()
    private val modeMenu = ModeMenu()
    private val trackMenu = TrackMenu()
    private val debugMenu = DebugMenu()
    private val collisionsMenu = CollisionsMenu()
    private val menu = listOf(startButton, modeMenu, trackMenu, collisionsMenu, debugMenu)

    private var currentMenu = 0

    private val lineWidth = 3f
    private val fontColor = Color(220, 200, 50)
    private val leftMargin = Game.screenWidth * 0.2f
    private val rightMargin = Game.screenWidth * 0.2f
    private val topMargin = Game.screenHeight * 0.2f
    private val bottomMargin = Game.screenHeight * 0.2f
    private val subMenuHeight = 36f
    private val subMenuItemHeight = 24f
    private val subMenuWidth = Game.screenWidth * 0.2f
    private val submenuFont = TrueTypeFont(Font(Font.SANS_SERIF, Font.BOLD, subMenuHeight.toInt()), true)
    private val itemFont = TrueTypeFont(Font(Font.SANS_SERIF, Font.BOLD, subMenuItemHeight.toInt()), true)
    private val listeners = mutableListOf<KeyListener>()

    override fun getID(): Int {
        return GameStateIds.getId(this::class)
    }

    @Throws(SlickException::class)
    override fun init(container: GameContainer, game: StateBasedGame) {
        val input = container.input
        GameSettings.instance.globalListeners.forEach( { input.addKeyListener(it) })
        KeyboardHelper.listeners.forEach( { input.addKeyListener(it) })

        listeners.add(SingleKeyAction({ -> menuChange(-1) }, KEY_UP))
        listeners.add(SingleKeyAction({ -> menuChange(+1) }, KEY_DOWN))
        listeners.add(SingleKeyAction({ -> selectionChange(-1) }, KEY_LEFT))
        listeners.add(SingleKeyAction({ -> selectionChange(+1) }, KEY_RIGHT))
        listeners.add(SingleKeyAction({ -> enter() }, KEY_RETURN))

        container.setTargetFrameRate(60)
    }

    private fun menuChange(delta: Int) {
        currentMenu += delta + menu.size
        currentMenu %= menu.size
        updateGameSettings()
    }

    private fun selectionChange(delta: Int) {
        menu[currentMenu].change(delta)
        updateGameSettings()
    }

    private fun enter() {
        updateGameSettings()
        startButton.enter()
    }

    @Throws(SlickException::class)
    override fun render(container: GameContainer, game: StateBasedGame, g: Graphics) {
        g.lineWidth = lineWidth
        g.color = fontColor

        menu.forEach { submenu ->
            val currentY = topMargin + menu.indexOf(submenu) * (subMenuHeight + 3 * lineWidth)
            drawSubMenu(g, submenu, currentY)

            drawSubmenuItems(g, submenu, currentY)
        }
    }

    private fun drawSubMenu(g: Graphics, submenu: SubMenu<*>, currentY: Float) {
        g.font = submenuFont
        g.drawString(submenu.title, leftMargin, currentY)
        if (currentMenu == menu.indexOf(submenu)) {
            g.drawLine(leftMargin, currentY + subMenuHeight + lineWidth,
                    leftMargin + subMenuWidth - 5 * lineWidth, currentY + subMenuHeight + lineWidth)
        }
    }

    private fun drawSubmenuItems(g: Graphics, submenu: SubMenu<*>, currentY: Float) {
        g.font = itemFont
        val itemY = currentY + subMenuHeight - subMenuItemHeight
        val items = submenu.items
        val itemWidth = (Game.screenWidth - leftMargin - rightMargin - subMenuWidth) / items.size
        items.forEach { item ->
            val currentX = leftMargin + subMenuWidth + submenu.items.indexOf(item) * itemWidth
            g.drawString(item.title, currentX, itemY)
            if (submenu.isCurrent(item)) {
                g.drawLine(currentX, itemY + subMenuItemHeight + lineWidth,
                        currentX + itemWidth - 5 * lineWidth, itemY + subMenuItemHeight + lineWidth)
            }
        }
    }

    @Throws(SlickException::class)
    override fun update(container: GameContainer, game: StateBasedGame, delta: Int) {

    }

    @Throws(SlickException::class)
    override fun leave(container: GameContainer?, game: StateBasedGame?) {
        super.leave(container, game)
        listeners.forEach { container!!.input.removeKeyListener(it) }
        GameSettings.instance.save()
    }

    private fun updateGameSettings() {
        GameSettings.instance.track = trackMenu.current
        GameSettings.instance.mode = modeMenu.current
        GameSettings.instance.debug = debugMenu.current
        GameSettings.instance.collisions = collisionsMenu.current
    }

    @Throws(SlickException::class)
    override fun enter(container: GameContainer?, game: StateBasedGame?) {
        super.enter(container, game)
        trackMenu.current = GameSettings.instance.track
        modeMenu.current = GameSettings.instance.mode
        debugMenu.current = GameSettings.instance.debug
        collisionsMenu.current = GameSettings.instance.collisions
        listeners.forEach { listener -> container!!.input.addKeyListener(listener) }

    }


}
